json/DataTypes/DecodeStudent.roc

```diff
@@ -0,0 +1,28 @@
+module [readStudents]
+
+import Student exposing [Student, Module]
+import Decoding exposing [JsonDecoder, list, field, string, number, bool, map2, map3, or]
+
+Students : List Student
+
+readStudents : JsonDecoder Students
+readStudents = \json ->
+    studentsField = \s -> (list readStudent) s
+    (field "students" studentsField) json
+
+readStudent : JsonDecoder Student
+readStudent = \json ->
+    nameField = field "name" string
+    modulesField = field "modules" (\m -> (list readModule) m)
+    currentGrade = field "currentGrade" number
+    finalGrade = field "finalGrade" number
+    currentStudent = map3 nameField modulesField currentGrade (\(name, mods, cg) -> CurrentStudent { name: name, modules: mods, currentGrade: cg })
+    graduatedStudent = map3 nameField modulesField finalGrade (\(name, mods, fg) -> GraduatedStudent { name: name, modules: mods, finalGrade: fg })
+    (or currentStudent graduatedStudent) json
+
+readModule : JsonDecoder Module
+readModule = \json ->
+    nameField = field "name" string
+    creditsField = field "credits" number
+    enrolledField = field "enrolled" bool
+    (map3 nameField creditsField enrolledField \(name, credits, enrolled) -> { name: name, credits: credits, enrolled: enrolled }) json
```

Comment:
I don't know if you want to get into record builder syntax in this chapter (or if it seems too advanced - also, the new version of this syntax just landed in the compiler, so it may not have existed when this part of the chapter was first written!) but it can make this function a lot more concise:

```roc
readModule : JsonDecoder Module
readModule = \json ->
    { map2 <-
        name: field "name" string,
        credits: field "credits" number,
        enrolled: field "enrolled" bool,
    } json
```

Reply:
Oh, although I think in order for this to work, `map2` would need the function it accepts to take 2 arguments instead of a tuple:

```diff
-map2 : JsonDecoder a, JsonDecoder b, ((a, b) -> c) -> JsonDecoder c
+map2 : JsonDecoder a, JsonDecoder b, (a, b -> c) -> JsonDecoder c
```

That's the `map2` signature the `{ foo <-` syntax sugar is designed to work with 😄